### PR TITLE
HIP backend

### PR DESCRIPTION
**Pull request by nanamiwang**
### How it works
Reply on PyHIP, a thin python wrapper for libamdhip64.so, to access ROCm HIP runtime API.
[https://github.com/jatinx/PyHIP](https://github.com/jatinx/PyHIP)

### How to run
PYTHONPATH=/path/to/PyHIP HIP=1 python3 examples/llama.py

### Speed test results on Ubuntu server 22.04, with RX7900XTX and ROCm 5.4.3.
Pytorch was running on CPU in this case.
Attempted running pytorch on GPU but failed due to this issue:
[https://discord.com/channels/1068976834382925865/1083520343169319042/1097708353536397372](https://discord.com/channels/1068976834382925865/1083520343169319042/1097708353536397372)
```
nanami@nanami-ai:~/tinygrad$ HIP=1 OPTLOCAL=1 JIT=1 python3.10 test/test_speed_v_torch.py
 add                               1x   1    0.00 ms (    0.00 GFLOPS     0.00 GB/s) in torch,    0.17 ms (    0.00 GFLOPS     0.00 GB/s) in tinygrad,   44.55x slower    0.00 MOPS    0.00 MB
 add                            1024x1024    0.31 ms (    3.37 GFLOPS    40.40 GB/s) in torch,    0.19 ms (    5.54 GFLOPS    66.43 GB/s) in tinygrad,    0.61x faster    1.05 MOPS   12.58 MB
 add                            4096x4096   13.45 ms (    1.25 GFLOPS    14.96 GB/s) in torch,    0.42 ms (   39.98 GFLOPS   479.72 GB/s) in tinygrad,    0.03x faster   16.78 MOPS  201.33 MB
.add_constant                   4096x4096   12.32 ms (    1.36 GFLOPS    10.89 GB/s) in torch,    0.30 ms (   56.03 GFLOPS   448.27 GB/s) in tinygrad,    0.02x faster   16.78 MOPS  134.22 MB
.add_sq                         4096x4096   40.41 ms (    1.25 GFLOPS     4.98 GB/s) in torch,    0.46 ms (  109.12 GFLOPS   436.48 GB/s) in tinygrad,    0.01x faster   50.33 MOPS  201.33 MB
.array_packing                  2048x2048    1.09 ms (    3.83 GFLOPS    30.65 GB/s) in torch,    0.20 ms (   20.90 GFLOPS   167.21 GB/s) in tinygrad,    0.18x faster    4.19 MOPS   33.55 MB
.conv bs: 32 chans:  4 ->  32                0.62 ms (  121.54 GFLOPS     7.71 GB/s) in torch,    0.18 ms (  412.10 GFLOPS    26.15 GB/s) in tinygrad,    0.29x faster   75.50 MOPS    4.79 MB
 conv bs: 32 chans: 16 ->  32                1.95 ms (  155.20 GFLOPS     3.38 GB/s) in torch,    0.20 ms ( 1520.79 GFLOPS    33.14 GB/s) in tinygrad,    0.10x faster  301.99 MOPS    6.58 MB
 conv bs: 32 chans: 64 ->  32                7.34 ms (  164.66 GFLOPS     1.87 GB/s) in torch,    0.27 ms ( 4443.94 GFLOPS    50.54 GB/s) in tinygrad,    0.04x faster 1207.96 MOPS   13.74 MB
.double_permute (64, 64, 64, 64)            16.50 ms (    1.02 GFLOPS     8.14 GB/s) in torch,    0.36 ms (   47.01 GFLOPS   376.12 GB/s) in tinygrad,    0.02x faster   16.78 MOPS  134.22 MB
.exp                            2048x2048    1.18 ms (    3.55 GFLOPS    28.38 GB/s) in torch,    0.21 ms (   20.00 GFLOPS   160.00 GB/s) in tinygrad,    0.18x faster    4.19 MOPS   33.55 MB
.gemm                           1024x1024   11.97 ms (  179.40 GFLOPS     1.05 GB/s) in torch,    0.36 ms ( 5942.29 GFLOPS    34.82 GB/s) in tinygrad,    0.03x faster 2147.48 MOPS   12.58 MB
.gemm                            256x 256    0.21 ms (  157.99 GFLOPS     3.70 GB/s) in torch,    0.22 ms (  152.48 GFLOPS     3.57 GB/s) in tinygrad,    1.04x slower   33.55 MOPS    0.79 MB
.gemm_unrolled                   512x 512    1.55 ms (  173.63 GFLOPS     2.03 GB/s) in torch,    0.21 ms ( 1288.91 GFLOPS    15.10 GB/s) in tinygrad,    0.13x faster  268.44 MOPS    3.15 MB
.gemm_unrolled_permute_l         512x 512    1.55 ms (  173.28 GFLOPS     2.03 GB/s) in torch,    0.24 ms ( 1128.33 GFLOPS    13.22 GB/s) in tinygrad,    0.15x faster  268.44 MOPS    3.15 MB
.gemm_unrolled_permute_lr        512x 512    1.54 ms (  174.24 GFLOPS     2.04 GB/s) in torch,    0.25 ms ( 1077.17 GFLOPS    12.62 GB/s) in tinygrad,    0.16x faster  268.44 MOPS    3.15 MB
.gemm_unrolled_permute_r         512x 512    1.53 ms (  175.22 GFLOPS     2.05 GB/s) in torch,    0.24 ms ( 1130.78 GFLOPS    13.25 GB/s) in tinygrad,    0.15x faster  268.44 MOPS    3.15 MB
.max                            4096x4096    3.08 ms (    5.45 GFLOPS    21.78 GB/s) in torch,    3.58 ms (    4.69 GFLOPS    18.75 GB/s) in tinygrad,    1.16x slower   16.78 MOPS   67.11 MB
.mul_sum                        4096x4096   16.38 ms (    2.05 GFLOPS     8.19 GB/s) in torch,    6.76 ms (    4.96 GFLOPS    19.85 GB/s) in tinygrad,    0.41x faster   33.55 MOPS  134.22 MB
.neg                            4096x4096   12.25 ms (    1.37 GFLOPS    10.95 GB/s) in torch,    0.29 ms (   57.10 GFLOPS   456.76 GB/s) in tinygrad,    0.02x faster   16.78 MOPS  134.22 MB
.conv bs:  1 chans: 12 ->  32                0.35 ms (  162.17 GFLOPS     4.17 GB/s) in torch,    0.20 ms (  289.27 GFLOPS     7.44 GB/s) in tinygrad,    0.56x faster   56.62 MOPS    1.46 MB
.partial_sum                    4096x4096    2.30 ms (    7.28 GFLOPS    29.12 GB/s) in torch,    0.36 ms (   46.91 GFLOPS   187.64 GB/s) in tinygrad,    0.16x faster   16.78 MOPS   67.11 MB
.permute                        1024x1024    0.54 ms (    1.93 GFLOPS    15.41 GB/s) in torch,    0.19 ms (    5.39 GFLOPS    43.14 GB/s) in tinygrad,    0.36x faster    1.05 MOPS    8.39 MB
 permute                        4096x4096   21.06 ms (    0.80 GFLOPS     6.37 GB/s) in torch,    0.57 ms (   29.42 GFLOPS   235.34 GB/s) in tinygrad,    0.03x faster   16.78 MOPS  134.22 MB
.pow                            2048x2048   11.93 ms (    0.35 GFLOPS     4.22 GB/s) in torch,    0.28 ms (   14.98 GFLOPS   179.77 GB/s) in tinygrad,    0.02x faster    4.19 MOPS   50.33 MB
.relu                           4096x4096   12.28 ms (    1.37 GFLOPS    10.93 GB/s) in torch,    0.29 ms (   56.96 GFLOPS   455.65 GB/s) in tinygrad,    0.02x faster   16.78 MOPS  134.22 MB
.sub                            4096x4096   13.71 ms (    1.22 GFLOPS    14.68 GB/s) in torch,    0.45 ms (   37.07 GFLOPS   444.80 GB/s) in tinygrad,    0.03x faster   16.78 MOPS  201.33 MB
.sum                            2048x2048    0.57 ms (    7.36 GFLOPS    29.43 GB/s) in torch,    1.04 ms (    4.04 GFLOPS    16.17 GB/s) in tinygrad,    1.82x slower    4.19 MOPS   16.78 MB
 sum                            4096x4096    2.24 ms (    7.48 GFLOPS    29.93 GB/s) in torch,    3.62 ms (    4.63 GFLOPS    18.52 GB/s) in tinygrad,    1.62x slower   16.78 MOPS   67.11 MB
.
----------------------------------------------------------------------
Ran 23 tests in 13.851s
```
### LLaMA can run,  though both weight loading and inference are much slower than OpenCL backend on the same RX7900XTX GPU.
```
nanami@nanami-ai:~/tinygrad$ HIP=1 OPTLOCAL=1 JIT=1 python3.10 examples/llama.py --timing --prompt "Hi"
using HIP backend
loading 288 ram used: 13.48 GB: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 323/323 [00:08<00:00, 38.28it/s]
loaded weights in 8457.51 ms, 13.48 GB loaded at 1.59 GB/s
Hi
ran model in 2251.48 ms
sync in 406.44 ms
,
ran model in 1852.02 ms
sync in 3.11 ms
 I
ran model in 676.95 ms
sync in 6.11 ms
'
ran model in 681.73 ms
sync in 2.13 ms
m
ran model in 679.61 ms
sync in 6.40 ms
 Jen
ran model in 689.06 ms
sync in 6.53 ms
ny
ran model in 684.28 ms
sync in 6.09 ms
,
ran model in 689.28 ms
sync in 2.17 ms
 and
ran model in 687.18 ms
sync in 6.42 ms
 I
ran model in 686.46 ms
sync in 6.20 ms
'
ran model in 690.50 ms
```
### TODO
Figure out why LLaMA inference is slow.

### CODE REVIEW

File: tinygrad/lazy.py [hunk 1]
```diff
@@ -292,7 +292,7 @@ def __init__(self) -> None:
   @functools.lru_cache(maxsize=None)  # this class is a singleton, pylint: disable=method-cache-max-size-none
   def __getitem__(self, x:str) -> Union[Interpreted, Compiled]: return [cls for cname, cls in inspect.getmembers(importlib.import_module(f'tinygrad.runtime.ops_{x.lower()}')) if (cname.lower() == x.lower() + "buffer") and x in self._buffers][0]
   def _default_device(self) -> str:
-    for device in ["METAL", "CUDA", "GPU"]:
+    for device in ["METAL", "CUDA", "GPU", "HIP"]:
```
Comment (geohot) [inline tinygrad/lazy.py:295]: Should GPU or HIP be the default if HIP is available? Currently it's GPU

**Comment by geohot** [comment]
llama might be slow because it's compiling kernels slower with HIP vs OpenCL?

**Comment by geohot** [comment]
Can "pip install -e .[hip]" install the stuff required for HIP?

File: tinygrad/lazy.py (hunk 1 above)
Comment (nanamiwang) [reply tinygrad/lazy.py:295]: Agreed, will fix this.

**Comment by nanamiwang** [comment]
> Can "pip install -e .[hip]" install the stuff required for HIP?
> 
> Actually, I'm reading this library, why depend on it? It should only be like 10 more lines to use ctypes, right?

Sure, the dependency is unnecessary, will remove it.

**Comment by nanamiwang** [comment]
> HIP

HIP compilation is indeed very slow. 8 kernels are compiled for every token, each of which takes around 80ms.
```
diff --git a/tinygrad/runtime/ops_hip.py b/tinygrad/runtime/ops_hip.py
index 6163e9a..f9a2591 100644
--- a/tinygrad/runtime/ops_hip.py
+++ b/tinygrad/runtime/ops_hip.py
@@ -21,7 +21,9 @@ class HIPProgram:
       if not binary:
         prog = hiprtc.hiprtcCreateProgram(prg, name, [], [])
         device_properties = hip.hipGetDeviceProperties(0)
-        hiprtc.hiprtcCompileProgram(prog, [f'--offload-arch={device_properties.gcnArchName}'])
+        from extra.helpers import Timing
+        with Timing():
+          hiprtc.hiprtcCompileProgram(prog, [f'--offload-arch={device_properties.gcnArchName}'])
         prg = hiprtc.hiprtcGetCode(prog)
     except hip.hipError as e:
       if DEBUG >= 3: print("FAILED TO BUILD", prg)
```

```
nanami@nanami-ai:~/tinygrad$ python3.10 examples/llama.py --prompt "How are you recently?" --count 64 --timing
using HIP backend
loading 288 ram used: 13.48 GB: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 323/323 [00:08<00:00, 39.09it/s]
loaded weights in 8272.18 ms, 13.48 GB loaded at 1.63 GB/s
How are you recently?
77.89 ms
83.95 ms
152.47 ms
80.73 ms
83.71 ms
78.52 ms
85.20 ms
75.86 ms
77.42 ms
75.44 ms
76.50 ms
87.99 ms
83.86 ms
128.39 ms
82.94 ms
83.75 ms
77.42 ms
88.42 ms
77.92 ms
81.68 ms
ran model in 1967.00 ms
75.64 ms
82.94 ms
77.38 ms
84.16 ms
76.19 ms
sync in 403.22 ms
 I
74.74 ms
82.37 ms
124.20 ms
76.82 ms
81.50 ms
78.03 ms
75.32 ms
80.46 ms
75.74 ms
76.47 ms
75.65 ms
75.69 ms
83.45 ms
127.06 ms
77.39 ms
81.88 ms
78.60 ms
85.18 ms
ran model in 1724.10 ms
sync in 2.38 ms
 trust
75.31 ms
75.49 ms
80.61 ms
76.26 ms
76.62 ms
75.36 ms
75.74 ms
84.68 ms
ran model in 821.01 ms
sync in 2.43 ms
 you
75.32 ms
84.20 ms
80.55 ms
76.41 ms
76.58 ms
75.77 ms
75.69 ms
86.09 ms
ran model in 832.75 ms
sync in 2.46 ms
 are
75.59 ms
75.62 ms
80.54 ms
76.41 ms
76.54 ms
75.66 ms
75.61 ms
87.53 ms
ran model in 824.96 ms
sync in 2.50 ms
 doing
75.41 ms
75.55 ms
80.61 ms
76.40 ms
76.59 ms
75.97 ms
75.92 ms
88.40 ms

```

**Comment by nanamiwang** [comment]
Use HIP as default for amd platforms.
Removed PyHIP dependency, add extra/hip_wrapper.py for HIP ctypes.
There are 598 lines of code added in hip_wrapper.py due to error handling and structures definition, specifically there is a huge structure definition for hipGetDeviceProperties. Is it ok?

**Comment by nanamiwang** [comment]
For the HIP ctypes, we may also need hipDeviceEnablePeerAccess and hipDeviceCanAccessPeer for P2P communication over PCIe in the future.

**Comment by geohot** [comment]
583 lines is not okay! How many do you actually need?

**Comment by nanamiwang** [comment]
231 non empty lines after removed unused functions, removed comments and simplify error check. As below:
```
try:
  _libhip = ctypes.cdll.LoadLibrary('libamdhip64.so')
except:
  raise OSError('cant find libamdhip64.so.')

try:
  _libhiprtc = ctypes.cdll.LoadLibrary('libhiprtc.so')
except:
  raise OSError('cant find libhiprtc.so.')

def hipCheckStatus(status):
  if status != 0: raise RuntimeError('HIP error %s' % status)

def hipDeviceSynchronize():
  hipCheckStatus(_libhip.hipDeviceSynchronize())

_libhip.hipEventCreate.restype = int
_libhip.hipEventCreate.argtypes = [ctypes.POINTER(ctypes.c_void_p)]

def hipEventCreate():
  ptr = ctypes.c_void_p()
  hipCheckStatus(_libhip.hipEventCreate(ctypes.byref(ptr)))
  return ptr


_libhip.hipEventRecord.restype = int
_libhip.hipEventRecord.argtypes = [ctypes.c_void_p, ctypes.c_void_p]

def hipEventRecord(event, stream=None):
  hipCheckStatus(_libhip.hipEventRecord(event, stream))


_libhip.hipEventSynchronize.restype = int
_libhip.hipEventSynchronize.argtypes = [ctypes.c_void_p]

def hipEventSynchronize(event):
  hipCheckStatus(_libhip.hipEventSynchronize(event))


_libhip.hipEventElapsedTime.restype = int
_libhip.hipEventElapsedTime.argtypes = [ctypes.POINTER(ctypes.c_float), ctypes.c_void_p, ctypes.c_void_p]

def hipEventElapsedTime(start, stop):
  t = ctypes.c_float()
  hipCheckStatus(_libhip.hipEventElapsedTime(ctypes.byref(t), start, stop))
  return t.value


_libhip.hipMalloc.restype = int
_libhip.hipMalloc.argtypes = [ctypes.POINTER(ctypes.c_void_p), ctypes.c_size_t]

def hipMalloc(count):
  ptr = ctypes.c_void_p()
  hipCheckStatus(_libhip.hipMalloc(ctypes.byref(ptr), count))
  return ptr


_libhip.hipMemcpyAsync.restype = int
_libhip.hipMemcpyAsync.argtypes = [ctypes.c_void_p, ctypes.c_void_p, ctypes.c_size_t, ctypes.c_int, ctypes.c_void_p]

def hipMemcpyAsync_htod(dst, src, count, stream):
  hipCheckStatus(_libhip.hipMemcpyAsync(dst, src, ctypes.c_size_t(count), 1, stream))


def hipMemcpyAsync_dtoh(dst, src, count, stream):
  hipCheckStatus(_libhip.hipMemcpyAsync(dst, src, ctypes.c_size_t(count), 2, stream))

class hipDeviceArch(ctypes.Structure):
  _fields_ = [
    ('hasGlobalInt32Atomics', ctypes.c_uint, 1),
    ('hasGlobalFloatAtomicExch', ctypes.c_uint, 1),
    ('hasSharedInt32Atomics', ctypes.c_uint, 1),
    ('hasSharedFloatAtomicExch', ctypes.c_uint, 1),
    ('hasFloatAtomicAdd', ctypes.c_uint, 1),
    ('hasGlobalInt64Atomics', ctypes.c_uint, 1),
    ('hasSharedInt64Atomics', ctypes.c_uint, 1),
    ('hasDoubles', ctypes.c_uint, 1),
    ('hasWarpVote', ctypes.c_uint, 1),
    ('hasWarpBallot', ctypes.c_uint, 1),
    ('hasWarpShuffle', ctypes.c_uint, 1),
    ('hasFunnelShift', ctypes.c_uint, 1),
    ('hasThreadFenceSystem', ctypes.c_uint, 1),
    ('hasSyncThreadsExt', ctypes.c_uint, 1),
    ('hasSurfaceFuncs', ctypes.c_uint, 1),
    ('has3dGrid', ctypes.c_uint, 1),
    ('hasDynamicParallelism', ctypes.c_uint, 1),
  ]

class hipDeviceProperties(ctypes.Structure):
  _fields_ = [
    ('_name', ctypes.c_char * 256),
    ('totalGlobalMem', ctypes.c_size_t),
    ('sharedMemPerBlock', ctypes.c_size_t),
    ('regsPerBlock', ctypes.c_int),
    ('warpSize', ctypes.c_int),
    ('maxThreadsPerBlock', ctypes.c_int),
    ('maxThreadsDim', ctypes.c_int * 3),
    ('maxGridSize', ctypes.c_int * 3),
    ('clockRate', ctypes.c_int),
    ('memoryClockRate', ctypes.c_int),
    ('memoryBusWidth', ctypes.c_int),
    ('totalConstMem', ctypes.c_size_t),
    ('major', ctypes.c_int),
    ('minor', ctypes.c_int),
    ('multiProcessorCount', ctypes.c_int),
    ('l2CacheSize', ctypes.c_int),
    ('maxThreadsPerMultiProcessor', ctypes.c_int),
    ('computeMode', ctypes.c_int),
    ('clockInstructionRate', ctypes.c_int),
    ('arch', hipDeviceArch),
    ('concurrentKernels', ctypes.c_int),
    ('pciDomainID', ctypes.c_int),
    ('pciBusID', ctypes.c_int),
    ('pciDeviceID', ctypes.c_int),
    ('maxSharedMemoryPerMultiProcessor', ctypes.c_size_t),
    ('isMultiGpuBoard', ctypes.c_int),
    ('canMapHostMemory', ctypes.c_int),
    ('gcnArch', ctypes.c_int),
    ('_gcnArchName', ctypes.c_char * 256),
    ('integrated', ctypes.c_int),
    ('cooperativeLaunch', ctypes.c_int),
    ('cooperativeMultiDeviceLaunch', ctypes.c_int),
    ('maxTexture1DLinear', ctypes.c_int),
    ('maxTexture1D', ctypes.c_int),
    ('maxTexture2D', ctypes.c_int * 2),
    ('maxTexture3D', ctypes.c_int * 3),
    ('hdpMemFlushCntl', ctypes.POINTER(ctypes.c_uint)),
    ('hdpRegFlushCntl', ctypes.POINTER(ctypes.c_uint)),
    ('memPitch', ctypes.c_size_t),
    ('textureAlignment', ctypes.c_size_t),
    ('texturePitchAlignment', ctypes.c_size_t),
    ('kernelExecTimeoutEnabled', ctypes.c_int),
    ('ECCEnabled', ctypes.c_int),
    ('tccDriver', ctypes.c_int),
    ('cooperativeMultiDeviceUnmatchedFunc', ctypes.c_int),
    ('cooperativeMultiDeviceUnmatchedGridDim', ctypes.c_int),
    ('cooperativeMultiDeviceUnmatchedBlockDim', ctypes.c_int),
    ('cooperativeMultiDeviceUnmatchedSharedMem', ctypes.c_int),
    ('isLargeBar', ctypes.c_int),
    ('asicRevision', ctypes.c_int),
    ('managedMemory', ctypes.c_int),
    ('directManagedMemAccessFromHost', ctypes.c_int),
    ('concurrentManagedAccess', ctypes.c_int),
    ('pageableMemoryAccess', ctypes.c_int),
    ('pageableMemoryAccessUsesHostPageTables', ctypes.c_int),
  ]

  @property
  def gcnArchName(self):
    return self._gcnArchName.decode('utf-8')


_libhip.hipGetDeviceProperties.restype = int
_libhip.hipGetDeviceProperties.argtypes = [ctypes.POINTER(hipDeviceProperties), ctypes.c_int]

def hipGetDeviceProperties(deviceId: int):
  device_properties = hipDeviceProperties()
  hipCheckStatus(_libhip.hipGetDeviceProperties(ctypes.pointer(device_properties), deviceId))
  return device_properties

_libhip.hipModuleLoadData.restype = int
_libhip.hipModuleLoadData.argtypes = [ctypes.POINTER(ctypes.c_void_p),  # Module
                                      ctypes.c_void_p]                 # Image


def hipModuleLoadData(data):
  module = ctypes.c_void_p()
  hipCheckStatus(_libhip.hipModuleLoadData(ctypes.byref(module), data))
  return module


_libhip.hipModuleGetFunction.restype = int
_libhip.hipModuleGetFunction.argtypes = [ctypes.POINTER(ctypes.c_void_p),  # Kernel
                                         ctypes.c_void_p,                    # Module
                                         ctypes.POINTER(ctypes.c_char)]      # kernel name


def hipModuleGetFunction(module, func_name):
  e_func_name = func_name.encode('utf-8')
  kernel = ctypes.c_void_p()
  hipCheckStatus(_libhip.hipModuleGetFunction(ctypes.byref(kernel), module, e_func_name))
  return kernel

_libhip.hipModuleLaunchKernel.restype = int
_libhip.hipModuleLaunchKernel.argtypes = [ctypes.c_void_p,                 # kernel
                                          ctypes.c_uint,                   # block x
                                          ctypes.c_uint,                   # block y
                                          ctypes.c_uint,                   # block z
                                          ctypes.c_uint,                   # thread x
                                          ctypes.c_uint,                   # thread y
                                          ctypes.c_uint,                   # thread z
                                          ctypes.c_uint,                   # shared mem
                                          ctypes.c_void_p,                 # stream
                                          # kernel params
                                          ctypes.POINTER(ctypes.c_void_p),
                                          ctypes.POINTER(ctypes.c_void_p)]  # extra


def hipModuleLaunchKernel(kernel, bx, by, bz, tx, ty, tz, shared, stream, struct):
  c_bx = ctypes.c_uint(bx)
  c_by = ctypes.c_uint(by)
  c_bz = ctypes.c_uint(bz)
  c_tx = ctypes.c_uint(tx)
  c_ty = ctypes.c_uint(ty)
  c_tz = ctypes.c_uint(tz)
  c_shared = ctypes.c_uint(shared)
  ctypes.sizeof(struct)
  hip_launch_param_buffer_ptr = ctypes.c_void_p(1)
  hip_launch_param_buffer_size = ctypes.c_void_p(2)
  hip_launch_param_buffer_end = ctypes.c_void_p(0)
  hip_launch_param_buffer_end = ctypes.c_void_p(3)
  size = ctypes.c_size_t(ctypes.sizeof(struct))
  p_size = ctypes.c_void_p(ctypes.addressof(size))
  p_struct = ctypes.c_void_p(ctypes.addressof(struct))
  config = (ctypes.c_void_p * 5)(hip_launch_param_buffer_ptr, p_struct,
                                 hip_launch_param_buffer_size, p_size, hip_launch_param_buffer_end)
  nullptr = ctypes.POINTER(ctypes.c_void_p)(ctypes.c_void_p(0))
  hipCheckStatus(_libhip.hipModuleLaunchKernel(kernel, c_bx, c_by, c_bz, c_tx, c_ty, c_tz, c_shared, stream, None, config))


_libhiprtc.hiprtcCreateProgram.restype = int
_libhiprtc.hiprtcCreateProgram.argtypes = [ctypes.POINTER(ctypes.c_void_p),  # hiprtcProgram
                                           ctypes.POINTER(ctypes.c_char),   # Source
                                           ctypes.POINTER(ctypes.c_char),   # Name
                                           ctypes.c_int,                    # numberOfHeaders
                                           ctypes.POINTER(ctypes.c_char_p),  # header
                                           ctypes.POINTER(ctypes.c_char_p)]  # headerNames


def hiprtcCreateProgram(source, name, header_names, header_sources):
  e_source = source.encode('utf-8')
  e_name = name.encode('utf-8')
  e_header_names = list()
  e_header_sources = list()
  for header_name in header_names:
    e_header_name = header_name.encode('utf-8')
    e_header_names.append(e_header_name)
  for header_source in header_sources:
    e_header_source = header_source.encode('utf-8')
    e_header_sources.append(e_header_source)

  prog = ctypes.c_void_p()
  c_header_names = (ctypes.c_char_p * len(e_header_names))()
  c_header_names[:] = e_header_names
  c_header_sources = (ctypes.c_char_p * len(e_header_sources))()
  c_header_sources[:] = e_header_sources
  hipCheckStatus(_libhiprtc.hiprtcCreateProgram(ctypes.byref(prog), e_source, e_name, len(e_header_names), c_header_sources, c_header_names))
  return prog


_libhiprtc.hiprtcCompileProgram.restype = int
_libhiprtc.hiprtcCompileProgram.argtypes = [ctypes.c_void_p,                 # hiprtcProgram
                                            ctypes.c_int,                    # num of options
                                            ctypes.POINTER(ctypes.c_char_p)]  # options

def hiprtcCompileProgram(prog, options):
  e_options = list()
  for option in options:
    e_options.append(option.encode('utf-8'))
  c_options = (ctypes.c_char_p * len(e_options))()
  c_options[:] = e_options
  hipCheckStatus(_libhiprtc.hiprtcCompileProgram(prog, len(c_options), c_options))


_libhiprtc.hiprtcGetCodeSize.restype = int
_libhiprtc.hiprtcGetCodeSize.argtypes = [ctypes.c_void_p,                 # hiprtcProgram
                                         ctypes.POINTER(ctypes.c_size_t)]  # Size of log
_libhiprtc.hiprtcGetCode.restype = int
_libhiprtc.hiprtcGetCode.argtypes = [ctypes.c_void_p,               # hiprtcProgram
                                     ctypes.POINTER(ctypes.c_char)]  # log


def hiprtcGetCode(prog):
  code_size = ctypes.c_size_t()
  status = _libhiprtc.hiprtcGetCodeSize(prog, ctypes.byref(code_size))
  hipCheckStatus(status)
  code = "0" * code_size.value
  e_code = code.encode('utf-8')
  hipCheckStatus(_libhiprtc.hiprtcGetCode(prog, e_code))
  return e_code
```

**Comment by guilt** [comment]
Hi there,

Isn't there a compiled prg cache available for most programs? Maybe the first time or two these kernels may take time, but one could introduce a pre-compile phase and ship binaries for inference.

**Comment by geohot** [comment]
Going to merge this, need HIP to access the WMMA.

Thanks! $200 bounty if you find a way to test HIP In CI.

**Comment by wozeparrot** [comment]
This does make it so that core tinygrad is importing from `extra/` do we want to move the bindings over or is it fine if we keep it in `extra/` since its not a default backend?

**Comment by geohot** [comment]
The "exec stub" wrapper is also in extra. The bindings code is ugly, it really belongs in a well tested and maintained pyhip library. But yea, I'm fine with it now since it's not default, but will have to do something before the 0.7 release.

**Comment by wozeparrot** [comment]
That makes sense, I may depend on this just to get p2p transfers to work as HIP does support them.

**Pull request merged**